include/seqan/seq_io/bam_sam.h

```diff
@@ -0,0 +1,521 @@
+inline void _writeRecord(TFile & file,
+                         TIdString const & meta,
+                         TSeqString const & seq,
+                         TQualString const & qual,
+                         Sam const & /* format */)
+{
+    if (SEQAN_UNLIKELY(!(context(file).headerWasWriten)))
+    {
+        BamHeader header;
+        fillHeader(header);
+
+        StringSet<CharString> contigNameStore;
+        appendValue(contigNameStore, "*");
+        NameStoreCache<StringSet<CharString> > contigNameStoreCache(contigNameStore);
+        BamIOContext<StringSet<CharString> > bamIOContext(contigNameStore, contigNameStoreCache);
+
+        file.context.bamIOContext = bamIOContext;
+        write(file.iter, header, context(file).bamIOContext, Sam());
+        file.context.headerWasWriten = true;
+    }
+
+    BamAlignmentRecord rec;
+    clear(rec);
+    rec.qName = meta;
+    rec.seq = seq;
+    rec.qual = qual;
+    rec.flag = BAM_FLAG_UNMAPPED;
+    
+    write(file.iter, rec, context(file).bamIOContext, Sam());
+}
+
+// ----------------------------------------------------------------------------
+// Function _writeRecord(BAM);
+// ----------------------------------------------------------------------------
+template <typename TFile, typename TIdString, typename TSeqString, typename TQualString>
+inline void _writeRecord(TFile & file,
+                         TIdString const & meta,
+                         TSeqString const & seq,
+                         TQualString const & qual,
+                         Bam const & /* format */)
+{
+    if (SEQAN_UNLIKELY(!(context(file).headerWasWriten)))
+    {
+        BamHeader header;
+        fillHeader(header);
+
+        StringSet<CharString> contigNameStore;
+        appendValue(contigNameStore, "*");
+        NameStoreCache<StringSet<CharString> > contigNameStoreCache(contigNameStore);
+        BamIOContext<StringSet<CharString> > bamIOContext(contigNameStore, contigNameStoreCache);
+
+        file.context.bamIOContext = bamIOContext;
+        write(file.iter, header, context(file).bamIOContext, Bam());
+        file.context.headerWasWriten = true;
+    }
+
+    BamAlignmentRecord rec;
+    clear(rec);
+    rec.qName = meta;
+    rec.seq = seq;
+    rec.qual = qual;
+    rec.flag = BAM_FLAG_UNMAPPED;
+    
+    write(file.iter, rec, context(file).bamIOContext, Bam());
+}
+
+// ----------------------------------------------------------------------------
+// Function writeRecord(SAM); Separate Qualities
+// ----------------------------------------------------------------------------
+template <typename TSpec, typename TIdString, typename TSeqString, typename TQualString>
+inline void writeRecord(FormattedFile<Fastq, Output, TSpec> & file,
+                        TIdString const & meta,
+                        TSeqString const & seq,
+                        TQualString const & qual,
+                        Sam const & tag)
```

Comment:
use `Sam const &` instead and then `Sam()` below so that this always gets optimized away